🐛 BUG: wrangler durableobject new-class migration fails for workers written in web-worker format

### Which Cloudflare product(s) does this pertain to?

Workers Runtime, Other

### What version(s) of the tool(s) are you using?

wrangler 3.27.0

### What version of Node are you using?

v18.19.0

### What operating system and version are you using?

Mac OS

### Describe the Bug

### Observed behavior
wrangler deploy does not work and errors with the code following error:
````
➜ wrangler deploy                      
 ⛅️ wrangler 3.27.0
-------------------
▲ [WARNING] The entrypoint src/requestHandler.ts has exports like an ES Module, but hasn't defined a default export like a module worker normally would. Building the worker using "service-worker" format...


Your worker has access to the following bindings:
- Durable Objects:
  - IP_STATS_DO: Ip***Do (defined in ./src/req***er.ts)
- KV Namespaces:
  - IP_STATS_KV_V3: 72d***3a07e
- Vars:
  - ENVIRONMENT: "staging"
Total Upload: 559.06 KiB / gzip: 104.16 KiB
 
✘ [ERROR] A request to the Cloudflare API (/accounts/***/workers/scripts/***) failed.

  Cannot apply new-class migration to class Ip***Do that is not exported by script [code: 10070]
````
### Expected behavior
The migration should work as expected

### Steps to reproduce
- Create a worker in service-worker format
- Create a migration in worker.toml
````
[durable_objects]
bindings = [
  { name = "IP_***_DO", class_name = "Ip***Do", script_name = "./src/ip***do.ts" }
]
[[migrations]]
tag = "v1"
new_classes = ["Ip***Do"]
````
- the error should occur error


### Please provide a link to a minimal reproduction

_No response_

### Please provide any relevant error logs

  Cannot apply new-class migration to class Ip***Do that is not exported by script [code: 10070]

--- a/.changeset/fix-do-service-worker-migrations-error.md
+++ b/.changeset/fix-do-service-worker-migrations-error.md
@@ -1,0 +1,5 @@
+---
+"wrangler": patch
+---
+
+fix: show a clear error when deploying a service-worker format Worker with Durable Object migrations instead of an opaque API error

--- a/.changeset/fix-do-service-worker-migrations-error.md
+++ b/.changeset/fix-do-service-worker-migrations-error.md
@@ -2,4 +2,4 @@
 "wrangler": patch
 ---
 
-fix: show a clear error when deploying a service-worker format Worker with Durable Object migrations instead of an opaque API error
+Show a clear error when deploying a service-worker format Worker with Durable Object migrations or bindings instead of an opaque API error

--- a/packages/wrangler/src/__tests__/deploy/bindings.test.ts
+++ b/packages/wrangler/src/__tests__/deploy/bindings.test.ts
@@ -1765,6 +1765,31 @@ describe("deploy", () => {
 					https://developers.cloudflare.com/workers/learning/migrating-to-module-workers/]
 				`);
 			});
+
+			it("should error when deploying service-worker worker with migrations", async ({
+				expect,
+			}) => {
+				writeWranglerConfig({
+					durable_objects: {
+						bindings: [],
+					},
+					migrations: [
+						{
+							tag: "v1",
+							new_classes: ["ExampleDurableObject"],
+						},
+					],
+				});
+				writeWorkerSource({ type: "sw" });
+				mockSubDomainRequest();
+
+				await expect(runWrangler("deploy index.js")).rejects
+					.toThrowErrorMatchingInlineSnapshot(`
+					[Error: Durable Object migrations require ES Module format Workers, but yours is being built as service-worker format. Migrations cannot be applied to service-worker format Workers.
+					Migrate your worker to ES Module syntax to use Durable Object migrations:
+					https://developers.cloudflare.com/workers/learning/migrating-to-module-workers/]
+				`);
+			});
 		});
 
 		describe("[services]", () => {

--- a/packages/wrangler/src/__tests__/deploy/bindings.test.ts
+++ b/packages/wrangler/src/__tests__/deploy/bindings.test.ts
@@ -1777,7 +1777,7 @@ describe("deploy", () => {
 				await expect(runWrangler("deploy index.js")).rejects
 					.toThrowErrorMatchingInlineSnapshot(`
 					[Error: Durable Object migrations require ES Module format Workers, but yours is being built as service-worker format. Migrations cannot be applied to service-worker format Workers.
-					Migrate your worker to ES Module syntax to use Durable Object migrations:
+					To use Durable Object migrations, deploy in ES Module format by adding a default export handler (e.g. "export default { fetch() {} }"), or remove "migrations" from your config if you don't need them. See:
 					https://developers.cloudflare.com/workers/learning/migrating-to-module-workers/]
 				`);
 			});

--- a/packages/wrangler/src/__tests__/deploy/bindings.test.ts
+++ b/packages/wrangler/src/__tests__/deploy/bindings.test.ts
@@ -96,12 +96,6 @@ describe("deploy", () => {
 	describe("bindings", () => {
 		it("should allow bindings with different names", async ({ expect }) => {
 			writeWranglerConfig({
-				migrations: [
-					{
-						tag: "v1",
-						new_classes: ["SomeDurableObject", "AnotherDurableObject"],
-					},
-				],
 				durable_objects: {
 					bindings: [
 						{
@@ -1770,9 +1764,6 @@ describe("deploy", () => {
 				expect,
 			}) => {
 				writeWranglerConfig({
-					durable_objects: {
-						bindings: [],
-					},
 					migrations: [
 						{
 							tag: "v1",

--- a/packages/wrangler/src/deployment-bundle/entry.ts
+++ b/packages/wrangler/src/deployment-bundle/entry.ts
@@ -133,7 +133,7 @@ export async function getEntry(
 			const errorMessage =
 				"Durable Object migrations require ES Module format Workers, but yours is being built as service-worker format. Migrations cannot be applied to service-worker format Workers.";
 			const migrateText =
-				"Migrate your worker to ES Module syntax to use Durable Object migrations:";
+				'To use Durable Object migrations, deploy in ES Module format by adding a default export handler (e.g. "export default { fetch() {} }"), or remove "migrations" from your config if you don\'t need them. See:';
 			const migrateUrl =
 				"https://developers.cloudflare.com/workers/learning/migrating-to-module-workers/";
 			throw new UserError(`${errorMessage}\n${migrateText}\n${migrateUrl}`, {

--- a/packages/wrangler/src/deployment-bundle/entry.ts
+++ b/packages/wrangler/src/deployment-bundle/entry.ts
@@ -118,7 +118,8 @@ export async function getEntry(
 			const errorMessage =
 				"You seem to be trying to use Durable Objects in a Worker written as a service-worker.";
 			const addScriptName = `You can use Durable Objects defined in other Workers by specifying a \`script_name\` in your ${configFileName(config.configPath)} file, where \`script_name\` is the name of the Worker that implements that Durable Object. For example:`;
-			const addScriptNameExamples = generateAddScriptNameExamples(localBindings);
+			const addScriptNameExamples =
+				generateAddScriptNameExamples(localBindings);
 			const migrateText =
 				"Alternatively, migrate your worker to ES Module syntax to implement a Durable Object in this Worker:";
 			const migrateUrl =
@@ -135,10 +136,10 @@ export async function getEntry(
 				"Migrate your worker to ES Module syntax to use Durable Object migrations:";
 			const migrateUrl =
 				"https://developers.cloudflare.com/workers/learning/migrating-to-module-workers/";
-			throw new UserError(
-				`${errorMessage}\n${migrateText}\n${migrateUrl}`,
-				{ telemetryMessage: "durable object migrations unsupported service worker" }
-			);
+			throw new UserError(`${errorMessage}\n${migrateText}\n${migrateUrl}`, {
+				telemetryMessage:
+					"durable object migrations unsupported service worker",
+			});
 		}
 	}
 

--- a/packages/wrangler/src/deployment-bundle/entry.ts
+++ b/packages/wrangler/src/deployment-bundle/entry.ts
@@ -113,32 +113,33 @@ export async function getEntry(
 
 	const { localBindings } = partitionDurableObjectBindings(config);
 
-	if (format === "service-worker" && localBindings.length > 0) {
-		const errorMessage =
-			"You seem to be trying to use Durable Objects in a Worker written as a service-worker.";
-		const addScriptName = `You can use Durable Objects defined in other Workers by specifying a \`script_name\` in your ${configFileName(config.configPath)} file, where \`script_name\` is the name of the Worker that implements that Durable Object. For example:`;
-		const addScriptNameExamples = generateAddScriptNameExamples(localBindings);
-		const migrateText =
-			"Alternatively, migrate your worker to ES Module syntax to implement a Durable Object in this Worker:";
-		const migrateUrl =
-			"https://developers.cloudflare.com/workers/learning/migrating-to-module-workers/";
-		throw new UserError(
-			`${errorMessage}\n${addScriptName}\n${addScriptNameExamples}\n${migrateText}\n${migrateUrl}`,
-			{ telemetryMessage: "tried to use DO with service worker" }
-		);
-	}
-
-	if (format === "service-worker" && config.migrations?.length) {
-		const errorMessage =
-			"Durable Object migrations require ES Module format Workers, but yours is being built as service-worker format. Migrations cannot be applied to service-worker format Workers.";
-		const migrateText =
-			"Migrate your worker to ES Module syntax to use Durable Object migrations:";
-		const migrateUrl =
-			"https://developers.cloudflare.com/workers/learning/migrating-to-module-workers/";
-		throw new UserError(
-			`${errorMessage}\n${migrateText}\n${migrateUrl}`,
-			{ telemetryMessage: "durable object migrations unsupported service worker" }
-		);
+	if (format === "service-worker") {
+		if (localBindings.length > 0) {
+			const errorMessage =
+				"You seem to be trying to use Durable Objects in a Worker written as a service-worker.";
+			const addScriptName = `You can use Durable Objects defined in other Workers by specifying a \`script_name\` in your ${configFileName(config.configPath)} file, where \`script_name\` is the name of the Worker that implements that Durable Object. For example:`;
+			const addScriptNameExamples = generateAddScriptNameExamples(localBindings);
+			const migrateText =
+				"Alternatively, migrate your worker to ES Module syntax to implement a Durable Object in this Worker:";
+			const migrateUrl =
+				"https://developers.cloudflare.com/workers/learning/migrating-to-module-workers/";
+			throw new UserError(
+				`${errorMessage}\n${addScriptName}\n${addScriptNameExamples}\n${migrateText}\n${migrateUrl}`,
+				{ telemetryMessage: "tried to use DO with service worker" }
+			);
+		}
+		if (config.migrations?.length) {
+			const errorMessage =
+				"Durable Object migrations require ES Module format Workers, but yours is being built as service-worker format. Migrations cannot be applied to service-worker format Workers.";
+			const migrateText =
+				"Migrate your worker to ES Module syntax to use Durable Object migrations:";
+			const migrateUrl =
+				"https://developers.cloudflare.com/workers/learning/migrating-to-module-workers/";
+			throw new UserError(
+				`${errorMessage}\n${migrateText}\n${migrateUrl}`,
+				{ telemetryMessage: "durable object migrations unsupported service worker" }
+			);
+		}
 	}
 
 	return {

--- a/packages/wrangler/src/deployment-bundle/entry.ts
+++ b/packages/wrangler/src/deployment-bundle/entry.ts
@@ -128,6 +128,19 @@ export async function getEntry(
 		);
 	}
 
+	if (format === "service-worker" && config.migrations?.length) {
+		const errorMessage =
+			"Durable Object migrations require ES Module format Workers, but yours is being built as service-worker format. Migrations cannot be applied to service-worker format Workers.";
+		const migrateText =
+			"Migrate your worker to ES Module syntax to use Durable Object migrations:";
+		const migrateUrl =
+			"https://developers.cloudflare.com/workers/learning/migrating-to-module-workers/";
+		throw new UserError(
+			`${errorMessage}\n${migrateText}\n${migrateUrl}`,
+			{ telemetryMessage: "durable object migrations unsupported service worker" }
+		);
+	}
+
 	return {
 		file: paths.absolutePath,
 		projectRoot,

